### PR TITLE
Use null object pattern to simplify dealing with guest users

### PR DESCRIPTION
**Pull request by johngallagher**
## What

A pattern to demonstrate the benefits by having a first class `GuestSocialUser` model to deal with guest users.

## Why

Because of the current complexity and error prone code.

### CODE REVIEW

File: _patterns/use-null-object-pattern-to-implement-guest-users.md [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+## Problems
+
+This has a number of issues:
+
+**1. Error prone.**
+
+Calling any method on `nil` results in a crash.
+
+So engineers must remember to use the safe navigator or use an `if current_user` guard clause around any code that uses `current_user`.
```
Comment (petershutt) [inline _patterns/use-null-object-pattern-to-implement-guest-users.md:18]: Also, new engineers must **learn** (sometimes the hard way) that `current_user` can be nil.

**Review by bockets — APPROVED** [review]

**Review by bhanubhakta — APPROVED** [review]

**Review by karlentwistle — APPROVED** [review]

**Review by petershutt — APPROVED** [review]

**Comment by karlentwistle** [comment]
This pattern blew my mind when I learnt about it at BathRuby 2015 https://www.youtube.com/watch?v=9lv2lBq6x4A

**Review by kxhitiz — APPROVED** [review]

**Comment by johngallagher** [comment]
On hold until @BiggerPockets/platform have created a PR with concrete examples of this pattern in effect for tracking.

**Review by RenatoCoelho-BiggerPockets — APPROVED** [review]

**Review by baerbradford — APPROVED** [review]

**Comment by johngallagher** [comment]
> On hold until @BiggerPockets/platform have created a PR with concrete examples of this pattern in effect for tracking.

This has been done and work is progressing with Guest users.

**Pull request merged**